src/routing_node.rs

```diff
@@ -1138,35 +1154,139 @@ impl RoutingNode {
+                                        }
+                                    }
+                                }
+                            }
+                            _ => {}
                         }
-                    }
-                    _ => None
+
+                    },
+                    _ => {}
                 }
-            }
-            _ => None,
+            },
+            None => {}
         }
+            
     }
 
-    fn handle_cache_put(&mut self, response: ExternalResponse) {
-        match response {
-            ExternalResponse::Get(data, _, _) => {
-                match data {
-                    Data::ImmutableData(ref immutable_data) => {
-                        self.data_cache.insert(immutable_data.name(), data.clone());
-                    }
-                    _ => {}
+    fn handle_cache_get(&mut self, message: &RoutingMessage) -> Option<Content> {
+        match self.data_cache {
+            Some(ref mut data_cache) => {
+                match message.content.clone() {
+                    Content::ExternalRequest(request) => {
+                        match request {
+                            ExternalRequest::Get(data_request, _) => {
+                                match data_request {
+                                    DataRequest::PlainData(data_name) => {
+                                        if self.cache_options.plain_data_caching_enabled() {
+                                            match data_cache.get(&data_name) {
+                                                Some(data) => {
+                                                    let response =
+                                                        ExternalResponse::Get(
+                                                            data.clone(),
+                                                            data_request,
+                                                            None);
+                                                    return Some(Content::ExternalResponse(response));
+                                                },
+                                                None => return None
+                                            }
+                                        }
+                                        return None;
+                                    }
+                                    DataRequest::StructuredData(data_name, version) => {
+                                        if self.cache_options.structured_data_caching_enabled() {
+                                            match data_cache.get(&data_name) {
+                                                Some(data) => {
+                                                    match *data {
+                                                        Data::StructuredData(ref structured_data) => {
+                                                            if structured_data.get_version() == version {
```

Comment:
It is worth noting that this is the important reason why safe_vault will not enable caching for structured data.  Routing is purely catering for generic uses here, and leaves it to the discretion of the user to assert that caching of structured data is sensible.